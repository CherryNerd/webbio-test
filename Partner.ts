import IPartner from "./Interfaces/IPartner";
import {Client} from "@googlemaps/google-maps-services-js";
import {headingDistanceTo} from 'geolocation-utils';

export default class Partner implements IPartner {

    address: { no: string; street: string; city: string; country: string };
    geoLocation: { lat: number; lon: number } = {lat: 0, lon: 0}; // Welcome to null island!
    name: string;
    distanceToWebbioOffice?: number;
    googleClient: Client;

    constructor(inputJson: IPartner) {
        Object.assign(this, inputJson);
        this.googleClient = new Client({});
    }

    async setup(): Promise<void> {
        let result;
        try {
            result = await this.googleClient.geocode({
                params: {
                    address: this.fullAddress,
                    key: process.env.GOOGLE_API_KEY
                },
            })

            const geometry = result.data.results[0].geometry.location;

            this.geoLocation.lat = geometry.lat;
            this.geoLocation.lon = geometry.lng;
        } catch (e) {
            if (e.isAxiosError) {
                throw {error: e.response.data};
            }
            throw {error: e};
        }
    }

    get fullAddress(): string {
        return `${this.address.street} ${this.address.no}, ${this.address.city}, ${this.address.country}`;
    }

    getDistanceInMeters(lat: number, lon: number): number {
        this.distanceToWebbioOffice = headingDistanceTo(this.geoLocation, {lat, lon}).distance;
        return Math.floor(this.distanceToWebbioOffice)
    }
}