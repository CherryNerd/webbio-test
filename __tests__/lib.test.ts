import {getPartnersWithinDistanceAndSortByName, loadOfficeLocation, loadPartners, setupPartners} from "../lib";
import Partner from "../Partner";
import * as fs from "fs/promises";
import {join} from 'path';

describe('loadPartners', () => {
    test('Should load the partners and return them as instances of Partner', async () => {
        const result = await loadPartners();

        expect(Array.isArray(result)).toBeTruthy();

        // @ts-ignore
        result.forEach(r => {
            expect(r).toBeInstanceOf(Partner)
            expect(r).toHaveProperty('fullAddress')
        });
    });

    test('Should return an error object if the JSON is invalid', async () => {
        const partnersJson = await fs.readFile(join(process.cwd(), 'partners.json'), 'utf-8');

        await fs.writeFile(join(process.cwd(), 'partners.json'), '{ "name": "partner_na', 'utf-8');

        try {
            await loadPartners();
        } catch (e) {
            expect(e).toBeInstanceOf(Error);
        }

        await fs.writeFile(join(process.cwd(), 'partners.json'), partnersJson, 'utf-8');
    });
});

describe('loadOfficeLocation', () => {
    test('Should load the partners and return them as instances of Partner', async () => {
        const result = await loadOfficeLocation();

        expect(result).toHaveProperty('lat');
        expect(result).toHaveProperty('lon');

    });

    test('Should return an error object if the JSON is invalid', async () => {
        const officeLocationJson = await fs.readFile(join(process.cwd(), 'OfficeLocation.json'), 'utf-8');

        await fs.writeFile(join(process.cwd(), 'OfficeLocation.json'), '{ "lat": 123', 'utf-8');

        try {
            await loadOfficeLocation();
        } catch (e) {
            expect(e).toBeInstanceOf(Error);
        }

        await fs.writeFile(join(process.cwd(), 'OfficeLocation.json'), officeLocationJson, 'utf-8');
    });
});

describe('setupPartners', () => {
    let partners;

    beforeEach(async () => {
        partners = await loadPartners();
    })

    test('Should call the setup function on each partner', async () => {
        const mock = {
            geocode: jest.fn().mockImplementation(async () => ({
                data: {
                    results: [{
                        geometry: {
                            location: {
                                lat: 12345,
                                lng: 67890,
                            }
                        }
                    }]
                }
            }))
        };

        partners.forEach(p => p.googleClient = mock);

        await setupPartners(partners);

        expect(mock.geocode).toHaveBeenCalledTimes(partners.length);

        partners.forEach(p => {
            expect(p.geoLocation).toHaveProperty('lat', 12345);
            expect(p.geoLocation).toHaveProperty('lon', 67890);
        });
    });
});

describe('getPartnersWithinDistance', () => {
    let partners;

    beforeEach(async () => {
        partners = await loadPartners();
        await setupPartners(partners);
    })

    test('Should throw an error if the office location file can\'t be read', async () => {
        const officeLocationJson = await fs.readFile(join(process.cwd(), 'OfficeLocation.json'), 'utf-8');

        await fs.writeFile(join(process.cwd(), 'OfficeLocation.json'), '{ "lat": 123', 'utf-8');

        try {
            await loadOfficeLocation();
        } catch (e) {
            expect(e).toBeInstanceOf(Error);
        }

        await fs.writeFile(join(process.cwd(), 'OfficeLocation.json'), officeLocationJson, 'utf-8');
    })

    test('Should filter partners within a certain amount of meters', async () => {
        expect(await getPartnersWithinDistanceAndSortByName(partners, 10000)).toMatchSnapshot()
        expect(await getPartnersWithinDistanceAndSortByName(partners, 25000)).toMatchSnapshot()
        expect(await getPartnersWithinDistanceAndSortByName(partners, 50000)).toMatchSnapshot()
        expect(await getPartnersWithinDistanceAndSortByName(partners, 75000)).toMatchSnapshot()
    });

});