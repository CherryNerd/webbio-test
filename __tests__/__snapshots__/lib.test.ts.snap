// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPartnersWithinDistance Should filter partners within a certain amount of meters 1`] = `
Array [
  Partner {
    "address": Object {
      "city": "Nijmegen",
      "country": "The Netherlands",
      "no": "149",
      "street": "Venusstraat",
    },
    "distanceToWebbioOffice": 3479.94616785502,
    "geoLocation": Object {
      "lat": 51.8364451,
      "lon": 5.8280268,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Alpite",
  },
  Partner {
    "address": Object {
      "city": "Nijmegen",
      "country": "The Netherlands",
      "no": "164",
      "street": "Waldeck Pyrmontsingel",
    },
    "distanceToWebbioOffice": 837.1351689415125,
    "geoLocation": Object {
      "lat": 51.8406408,
      "lon": 5.866327699999999,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Wiredix",
  },
]
`;

exports[`getPartnersWithinDistance Should filter partners within a certain amount of meters 2`] = `
Array [
  Partner {
    "address": Object {
      "city": "Nijmegen",
      "country": "The Netherlands",
      "no": "149",
      "street": "Venusstraat",
    },
    "distanceToWebbioOffice": 3479.94616785502,
    "geoLocation": Object {
      "lat": 51.8364451,
      "lon": 5.8280268,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Alpite",
  },
  Partner {
    "address": Object {
      "city": "Nijmegen",
      "country": "The Netherlands",
      "no": "164",
      "street": "Waldeck Pyrmontsingel",
    },
    "distanceToWebbioOffice": 837.1351689415125,
    "geoLocation": Object {
      "lat": 51.8406408,
      "lon": 5.866327699999999,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Wiredix",
  },
]
`;

exports[`getPartnersWithinDistance Should filter partners within a certain amount of meters 3`] = `
Array [
  Partner {
    "address": Object {
      "city": "Nijmegen",
      "country": "The Netherlands",
      "no": "149",
      "street": "Venusstraat",
    },
    "distanceToWebbioOffice": 3479.94616785502,
    "geoLocation": Object {
      "lat": 51.8364451,
      "lon": 5.8280268,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Alpite",
  },
  Partner {
    "address": Object {
      "city": "Leusden",
      "country": "The Netherlands",
      "no": "84",
      "street": "Vogelwikke",
    },
    "distanceToWebbioOffice": 44822.854480770475,
    "geoLocation": Object {
      "lat": 52.1390782,
      "lon": 5.4413952,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Angelico",
  },
  Partner {
    "address": Object {
      "city": "Nijmegen",
      "country": "The Netherlands",
      "no": "164",
      "street": "Waldeck Pyrmontsingel",
    },
    "distanceToWebbioOffice": 837.1351689415125,
    "geoLocation": Object {
      "lat": 51.8406408,
      "lon": 5.866327699999999,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Wiredix",
  },
]
`;

exports[`getPartnersWithinDistance Should filter partners within a certain amount of meters 4`] = `
Array [
  Partner {
    "address": Object {
      "city": "Nijmegen",
      "country": "The Netherlands",
      "no": "149",
      "street": "Venusstraat",
    },
    "distanceToWebbioOffice": 3479.94616785502,
    "geoLocation": Object {
      "lat": 51.8364451,
      "lon": 5.8280268,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Alpite",
  },
  Partner {
    "address": Object {
      "city": "Leusden",
      "country": "The Netherlands",
      "no": "84",
      "street": "Vogelwikke",
    },
    "distanceToWebbioOffice": 44822.854480770475,
    "geoLocation": Object {
      "lat": 52.1390782,
      "lon": 5.4413952,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Angelico",
  },
  Partner {
    "address": Object {
      "city": "Utrecht",
      "country": "The Netherlands",
      "no": "39",
      "street": "Zwarte Woud",
    },
    "distanceToWebbioOffice": 56332.42878904909,
    "geoLocation": Object {
      "lat": 52.0643559,
      "lon": 5.1427641,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Herbtechs",
  },
  Partner {
    "address": Object {
      "city": "Tilburg",
      "country": "The Netherlands",
      "no": "171",
      "street": "Treubstraat",
    },
    "distanceToWebbioOffice": 64605.55611778703,
    "geoLocation": Object {
      "lat": 51.5660039,
      "lon": 5.052448,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Phantomedia",
  },
  Partner {
    "address": Object {
      "city": "Lichtenvoorde",
      "country": "The Netherlands",
      "no": "9",
      "street": "Tongerlosestraat",
    },
    "distanceToWebbioOffice": 50624.59162679687,
    "geoLocation": Object {
      "lat": 51.98458189999999,
      "lon": 6.5770975,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Shadoworks",
  },
  Partner {
    "address": Object {
      "city": "Enschede",
      "country": "The Netherlands",
      "no": "185",
      "street": "Hagmolenbeekweg",
    },
    "distanceToWebbioOffice": 73361.44989973035,
    "geoLocation": Object {
      "lat": 52.16925149999999,
      "lon": 6.805305,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Wavetech",
  },
  Partner {
    "address": Object {
      "city": "Hilversum",
      "country": "The Netherlands",
      "no": "112",
      "street": "Van der Duyn van Maasdamstraat",
    },
    "distanceToWebbioOffice": 64648.627331149335,
    "geoLocation": Object {
      "lat": 52.2140381,
      "lon": 5.1569519,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Wavigations",
  },
  Partner {
    "address": Object {
      "city": "Nijmegen",
      "country": "The Netherlands",
      "no": "164",
      "street": "Waldeck Pyrmontsingel",
    },
    "distanceToWebbioOffice": 837.1351689415125,
    "geoLocation": Object {
      "lat": 51.8406408,
      "lon": 5.866327699999999,
    },
    "googleClient": Client {
      "axiosInstance": [Function],
    },
    "name": "Wiredix",
  },
]
`;
